quipucords/scanner/normalizer.py

```diff
@@ -0,0 +1,169 @@
+    source_type = None
+
+    def __init__(self, raw_facts: dict, server_id: str = None) -> None:
+        """Create a normalizer instance.
+
+        :param raw_facts: dict of raw facts from inspection phase
+        :param server_id: server_id of quipucords server that collected these
+        raw facts.
+        """
+        self.raw_facts = raw_facts
+        self._metadata = {}
+        self.server_id = server_id or get_server_id()
+
+    @property
+    def facts(self):
+        """Return normalized facts."""
+        assert hasattr(
+            self, "_normalized"
+        ), "Call 'normalize()' before asking for facts."
+        return self._normalized
+
+    def normalize(self):
+        """Normalize and validate raw facts.
+
+        :return: dict of normalized facts
+        """
+        if hasattr(self, "_normalized"):
+            return self._normalized
+        self._normalized = {}
+        for fact_name in self.fields.keys():
+            fact_value, raw_fact_keys = self._normalize_fact(fact_name)
+            self._normalized[fact_name] = fact_value
+            self._metadata[fact_name] = {
+                "raw_fact_keys": raw_fact_keys,
+                "source_type": self.source_type,
+                "server_id": self.server_id,
+            }
+        return self._normalized
+
+    def _normalize_fact(self, fact_name):
+        """
+        Normalize given fact.
+
+        Blind exceptions are used to ensure issues with normalization or validation
+        functions don't affect other facts.
+
+        Returns the normalized fact and a list a raw facts used for normalization.
+        """
+        field: FactMapper = self.fields[fact_name]
+        kwargs = {dep: self._normalized[dep] for dep in field.dependencies}
+        if len(field.raw_facts) == 1:
+            raw_fact_key = field.raw_facts[0]
+            raw_value = self.raw_facts.get(raw_fact_key)
+            args = (raw_value,)
+        else:
+            args = ()
+            kwargs.update(**{raw: self.raw_facts.get(raw) for raw in field.raw_facts})
+        try:
+            normalized_fact = field.normalizer(*args, **kwargs)
+        except Exception:  # where's the blind exception error, ruff?
+            logger.exception("Unexpected error during '%s' normalization.", fact_name)
+            return None, None
+        try:
+            for validator in field.validators:
+                if not validator(normalized_fact):
+                    logger.error(
+                        "'{fact_name}' is invalid (value={normalized_fact})",
+                        fact_name=fact_name,
+                        normalized_fact=normalized_fact,
+                    )
+                    return None, None
+        except Exception:
+            logger.exception("Unexpected error during '%s' validation.", fact_name)
+            return None, None
+        # for now, assume all facts contributed
+        return normalized_fact, field.raw_facts
```

Comment:
TODO: implement the API for specifying which raw facts/dependencies) were used for a fact